### DISCUSSION TITLE
Add "previousValue" parameter to @PostUpdate @PreUpdate handler

### DISCUSSION THREAD

**Discussion by trajano** [jakartaee/persistence#641] [Ideas]
I am pretty sure this can't be done easily because we only have references and can't simply just "copy" the previous value first for @PreUpdate and @PostUpdate

That is saying something like

```java
@PreUpdate
void aboutToUpdate(Entity out, Entity in) {}

@PostUpdate
void updated(Entity out, Entity in) {}
```

The idea I had that if a second parameter is provided namely `in` then it's the object that was about to be updated but detached.  The thing is if it was still pointing to the same values as the `out` entity they'd be the same values.

So for someone to work around that `in` may have a `transient` variable containing the older values which will be dropped off the `out` parameter.  Then they can do some manipulations.

Anyway just a random idea.

**Comment by lprimak** [comment]
Upvoted

**Comment by gavinking** [comment]
My knee-jerk is that this is really a *lot* harder than it sounds, and has nontrivial implications for performance.

For example, what is the expectation for associations?

Note that Hibernate offers something like this via the `Interceptor` API, but there the previous state is provided in the form of a flattened and deconstructed `state` array, which circumvents any problems with association access.

**Reply by trajano** [reply]
yeah I agree, I am not really sure the best way of handling this.  Except for the point that `in` is detached and so won't have any associations unless already "fetched" earlier.

The use case is simply to provide a way of determining some values that are on the top level has been updated I guess a good common one would be some LOB of JSON and we simply want to compare that the `in` json is the same as the `out` json.

**Reply by lprimak** [reply]
In my use case, only basic values can be supported. Collections can be safely ignored and not supported. 
Since weaving already keeps previous values, I would think this could be easily implemented in both hibernate and eclipselonk

**Reply by lprimak** [reply]
On the other hand, I am just guessing about the internals. Perhaps just Hibernate Interceptor needs to be standardized. Perhaps a CDI interceptor can be made to work?

What do you think Gavin?

**Reply by gavinking** [reply]
I would not attempt to standardize `Interceptor` itself, but I think we could consider allowing something like the following:

```java
@PreUpdate
void onUpdate(Entity entity, AttributeUpdate<Entity,?>... attributeUpdates) { ... }
```

where `AttributeUpdate` is defined by:

```java
interface AttributeUpdate<E,T> {
    SingularAttribute<? super E,T> attribute();
    T oldValue();
    T newValue();
}
```

For generic listeners I guess you would write something like:

```java
@PreUpdate
<E> void onUpdate(E entity, AttributeUpdate<E,?>... attributeUpdates) { ... }
```

We would need to think very carefully about how this relates to the new `EntityAgent` stuff. I assume that these things would be `AttributeInsert`, `AttributeUpdate`, and `AttributeDelete`, corresponding to the `@PreInsert`, `@PreUpdate`, `@PreDelete` event types (not to `@PrePersist`, `@PreMerge`, and `@PreRemove`). But for `EntityAgent.update()`, old values are not known. Anyway, that's more of a detail I guess.

**Reply by gavinking** [reply]
> But for `EntityAgent.update()`, old values are not known. Anyway, that's more of a detail I guess.

And in `@PreUpsert`, we're not even sure if they're `AttributeUpdate`s or `AttributeInsert`s. But these details aren't showstoppers. At worst we can just define an `AttributeOverwite` or whatever.

**Reply by lprimak** [reply]
This would definitely work for my use case. Thank you!
